### PR TITLE
Fix errors reported by phpstan level 3.

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

File: src/Database/ExpressionInterface.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ public function sql(ValueBinder $generator): string;
      * being iterated.
      *
      * @param callable $visitor The callable to apply to all nodes.
-     * @return $this
+     * @return $this|void
```
Comment (dereuromark) [inline src/Database/ExpressionInterface.php]: This cannot be. Didnt I remove this as the void is only inside the callable etc?

File: src/Console/Shell.php [hunk 2]
```diff
@@ -37,7 +37,7 @@
  * Is the equivalent of Cake\Controller\Controller on the command line.
  *
  * @deprecated 3.6.0 ShellDispatcher and Shell will be removed in 5.0
- * @method int|bool|null main(...$args)
+ * @method int|bool|void main(...$args)
```
[the remark this answers is not in the document]
Comment (ADmad) [reply src/Console/Shell.php]: I know but shells are deprecated anyway and I don't want to got about updating `main()` of all shells to ruturn null.

[the remark this answers is not in the document]
Comment (dereuromark) [reply src/Console/Shell.php]: better to silence this then IMO. as this is dead wrong and will cause all inherit to declare the same wrong thing.

File: src/Database/ExpressionInterface.php (hunk 1 above)
Comment (ADmad) [reply src/Database/ExpressionInterface.php]: `traverse()` method of all expression classes which implement this interface return void. Only `Query::traverse()` returns `$this`.

Comment (dereuromark) [reply src/Database/ExpressionInterface.php]: Couldnt all return this then? for a clean api?

File: src/Console/Shell.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ADmad) [reply src/Console/Shell.php]: Okay, I'll revert this and ignore the error then.

File: src/Database/ExpressionInterface.php (hunk 1 above)
Comment (ADmad) [reply src/Database/ExpressionInterface.php]: Fine by me if @lorenzo / @markstory agree.

File: phpstan.neon [hunk 3]
```diff
@@ -27,6 +27,7 @@ parameters:
         -
             message: '#Binary operation "-" between (.)* results in an error#'
             path: 'src/I18n/RelativeTimeFormatter.php'
+        - '#Static property Cake\\Chronos\\(Date|Chronos|MutableDate|MutableDateTime)::\$diffFormatter \(Cake\\Chronos\\DifferenceFormatter\) does not accept Cake\\I18n\\RelativeTimeFormatter#'
```
Comment (ADmad) [inline phpstan.neon]: A proper fix for this would be to add an interface for diff formatter to Chronos so that it property can be type hinted with the interface instead of concrete class.

File: phpstan.neon [hunk 4]
```diff
@@ -27,6 +27,17 @@ parameters:
...
+        - '#Return type \(void\) of method Cake\\Shell\\[A-Za-z]+Shell::main\(\) should be compatible with return type \(bool|int|null\) of method Cake\\Console\\Shell::main\(\)#'
+        - '#Property Cake\\TestSuite\\Fixture\\TestFixture::\$_schema \(Cake\\Database\\Schema\\SqlGeneratorInterface&Cake\\Database\\Schema\\TableSchemaInterface\) does not accept Cake\\Database\\Schema\\TableSchemaInterface#'
+        - '#Method Cake\\ORM\\Query::find\(\) should return \$this\(Cake\\Datasource\\QueryInterface\) but returns Cake\\ORM\\Query#'
+        -
+            message: '#Offset mixed does not exist on array\(\)|array\(2 => mixed\)|array\(string, string\)#'
+            path: 'src/Database/Dialect/SqlserverDialectTrait.php'
+        -
+            message: '#Offset 0 does not exist on array\(\)#'
+            path: 'src/Database/Dialect/SqlserverDialectTrait.php'
+        - '#Return type \(null\) of method Cake\\Collection\\Iterator\\NoChildrenIterator::getChildren\(\) should be compatible with return type \(RecursiveIterator\) of method RecursiveIterator::getChildren\(\)#'
```
Comment (ADmad) [inline phpstan.neon]: @lorenzo What would be a proper fix for this?

**Comment by ADmad** [comment]
There shouldn't be any errors reported at phpstan level 3 now :muscle:.

File: src/Cache/Engine/FileEngine.php [hunk 5]
```diff
@@ -319,25 +319,31 @@ protected function _clearDirectory(string $path): void
...
+     * @return false
      * @throws \LogicException
      */
-    public function decrement(string $key, int $offset = 1): void
+    public function decrement(string $key, int $offset = 1)
     {
         throw new LogicException('Files cannot be atomically decremented.');
+
+        // phpcs:ignore
+        return false;
```
Comment (markstory) [inline src/Cache/Engine/FileEngine.php]: Why is this necessary? Seems like a phpstan issue.

File: src/Database/ExpressionInterface.php (hunk 1 above)
Comment (markstory) [reply src/Database/ExpressionInterface.php]: Going from void -> this is safe.

File: src/Cache/Engine/FileEngine.php (hunk 5 above)
Comment (ADmad) [reply src/Cache/Engine/FileEngine.php]: Docblock of CacheEngineInterface::decrement() says it should return a boolean or int, so phpstan is unhappy if there's no retuy value.

But yes logically return value should not be necessary due to the exception. I can revert and ignore the error instead.

Comment (dereuromark) [reply src/Cache/Engine/FileEngine.php]: We should Report this as phpstan bug. It clearly is.

Comment (ADmad) [reply src/Cache/Engine/FileEngine.php]: Turned out the only problem was the `void` return type hint.
